fvm/src/syscalls/mod.rs

```diff
@@ -29,15 +30,63 @@ pub struct InvocationData<K> {
     /// The last-seen syscall error. This error is considered the abort "cause" if an actor aborts
     /// after receiving this error without calling any other syscalls.
     pub last_error: Option<backtrace::Cause>,
+
+    /// This snapshot is used to track changes in available_gas during syscall invocations.
+    /// The snapshot gets taken when execution exits WASM _after_ charging gas for any newly incurred fuel costs.
+    /// When execution moves back into WASM, we consume fuel for the delta between the snapshot and the new gas_available value.
+    pub gas_available_snapshot: i64,
+
+    /// This snapshot is used to track changes in fuel_consumed during WASM execution.
+    /// The snapshot gets taken when execution enters WASM _after_ consuming fuel for any syscall gas consumption.
+    /// When execution exits WASM, we charge gas for the delta between the new fuel_consumed value and the snapshot.
+    pub exec_units_consumed_snapshot: u64,
 }
 
-impl<K> InvocationData<K> {
+impl<K: Kernel> InvocationData<K> {
     pub(crate) fn new(kernel: K) -> Self {
+        let gas_available = kernel.gas_available();
         Self {
             kernel,
             last_error: None,
+            gas_available_snapshot: gas_available,
+            exec_units_consumed_snapshot: 0,
         }
     }
+
+    /// This method:
+    /// 1) calculates the available_gas delta from the previous snapshot,
+    /// 2) converts this to the corresponding amount of exec_units
+    /// 3) updates the available_gas and exec_units_consumed snapshots. The exec_units_consumed_snapshot is optimistically updated, assuming the value calculated in 2 will be consumed
+    /// 4) returns the value calculated in 2) for its caller to actually consume that exec_units_consumed
+    pub(crate) fn calculate_exec_units_for_gas(&mut self) -> Result<u64> {
+        let gas_available = self.kernel.gas_available();
+        let gas_used = self.gas_available_snapshot - gas_available;
+        if gas_used < 0 {
+            return Err(ExecutionError::Syscall(SyscallError(
+                String::from("used more gas than available"),
+                fvm_shared::error::ErrorNumber::IllegalOperation,
+            )));
```

Comment:
On reflection, I don't think this is correct. Technically we can refund gas (and this _definitely_ isn't a syscall error).

I _think_ we need to add fuel (which we can do). And I _don't_ think that causes any other problems.... but we'll have to be careful.

Reply:
> it means the gas_available somehow increased as execution went on, so I'm p sure it can only be programmer error

Nope. We have negative gas charges in some very special cases.